README.md
[![Build Status](https://api.travis-ci.org/zambal/eml.svg?branch=master)](https://travis-ci.org/zambal/eml)

# Eml

## Markup for developers

### What is it?
Eml makes markup a first class citizen in Elixir. It provides a flexible and
modular toolkit for generating, parsing and manipulating markup. It's main focus
is html, but other markup languages could be implemented as well.

To start off:

This piece of code
```elixir
use Eml.HTML

name = "Vincent"
age  = 36

div class: "person" do
  div do
    span "name: "
    span name
  end
  div do
    span "age: "
    span age
  end
end |> Eml.compile
```

produces
```html
<div class='person'>
  <div>
    <span>name: </span>
    <span>Vincent</span>
  </div>
  <div>
    <span>age: </span>
    <span>36</span>
  </div>
</div>
```

### Why?
Most templating libraries are build around the idea of interpreting strings that
can contain embeded code. This code is mostly used for implementing view logic
in the template. You could say that these libraries are making code a first
class citizen in template strings. As long as the view logic is simple this
works pretty well, but with more complex views this can become quite messy. Eml
takes this idea inside out and makes the markup that you normally would write as
a string a first class citizen of the programming language, allowing you to
compose and organize markup and view logic with all the power of Elixir.

Please read on for a walkthrough that tries to cover most of Eml's features.


### Walkthrough

- [Intro](#intro)
- [Compiling](#compiling)
- [Parsing](#parsing)
- [Compiling and templates](#compiling-and-templates)
- [Components and fragments](#components-and-fragments)
- [Unpacking](#unpacking)
- [Querying eml](#querying-eml)
- [Transforming eml](#transforming-eml)
- [Encoding data in Eml](#encoding-data-in-eml)
- [Notes](#notes)

#### Intro

```elixir
iex> use Eml
nil
iex> use Eml.HTML
nil
```

By invoking `use Eml`, some macro's are imported into the current scope and core
API modules are aliased. `use Eml.HTML` imports all generated html element
macros from `Eml.HTML` into the current scope. The element macros just translate
to a call to `%Eml.Element{...}`, a struct that is the actual representation of
elements, so they can even be used inside a match.
```elixir
iex> div 42
#div<42>
```
Here we created a `div` element with `42` as it contents.

The element macro's in Eml try to be clever about the type of arguments that get
passed. For example, if the first argument is a Keyword list, it will be
interpreted as attributes, otherwise as content.
```elixir
iex> div id: "some-id"
#div<%{id: "some-id"}>

iex> div "some content"
#div<"some content">

iex> div do
...>   "some content"
...> end
#div<["some content"]>

iex> div [id: "some-id"], "some content"
#div<%{id: "some-id"} "some content">

iex> div id: "some-id" do
...>   "some content"
...> end
#div<%{id: "some-id"} ["some content"]>

iex> %Element{tag: :div, attrs: %{id: "some-id"}, content: "some content"}
#div<%{id: "some-id"} "some content">
```

Note that attributes are stored internally as a map.


#### Compiling

Contents can be compiled to a string by calling `Eml.compile`. Eml automatically
inserts a doctype declaration when the html element is the root.
```elixir
iex> html(body(div(42))) |> Eml.compile
"<!doctype html>\n<html><body><div>42</div></body>\n</html>"

iex> "text & more" |> div |> body |> html |> Eml.compile
"<!doctype html>\n<html><body><div>text &amp; more</div></body></html>"
```
As you can see, you can also use Elixir's pipe operator for creating markup.
However, using do blocks, as can be seen in the introductory example, is more
convenient most of the time.

#### Parsing

Eml's parser by default converts a string with html content into Eml content.
```elixir
iex> Eml.parse "<!doctype html>\n<html><head><meta charset='UTF-8'></head><body><div>42</div></body></html>"
[#html<[#head<[#meta<%{charset: "UTF-8"}>]>, #body<[#div<"42">]>]>]

iex> Eml.parse "<div class=\"content article\"><h1 class='title'>Title<h1><p class=\"paragraph\">blah &amp; blah</p></div>"
[#div<%{class: "content article"}
 [#h1<%{class: "title"}
  ["Title", #h1<[#p<%{class: "paragraph"} "blah & blah">]>]>]>]
```

The html parser is primarily written to parse html compiled by Eml, but it's
flexible enough to parse most html you throw at it. Most notable missing
features of the parser are attribute values without quotes and elements that are
not properly closed.


#### Compiling and templates

Compiling and templates can be used in situations where most content is static
and performance is critical, since its contents gets precompiled during compiletime.

Eml uses the assigns extension from `EEx` for parameterizing templates. See
the `EEx` docs for more info about them. The function that the template macro
defines accepts optionally any Dict compatible dictionary as argument for
binding values to assigns.

```elixir
iex> defmodule MyTemplates1 do
...>   use Eml
...>   use Eml.HTML
...>
...>   template example do
...>     div id: "example" do
...>       span @text
...>     end
...>   end
...> end
iex> MyTemplates.example text: "Example text"
{:safe, "<div id='example'><span>Example text</span></div>"}
```

Eml templates provides two ways of executing logic during runtime. By
providing assigns handlers to the optional `funs` dictionary, or by calling
external functions during runtime with the `&` operator.

```elixir
iex> defmodule MyTemplates2 do
...>   use Eml
...>   use Eml.HTML
...>
...>   template assigns_handler,
...>   text: &String.upcase/1 do
...>     div id: "example1" do
...>       span @text
...>     end
...>   end
...>
...>   template external_call do
...>     body &assigns_handler(text: @example_text)
...>   end
...> end
iex> MyTemplates.assigns_handler text: "Example text"
{:safe, "<div id='example'><span>EXAMPLE TEXT</span></div>"}
iex> MyTemplates.exernal_call example_text: "Example text"
{:safe, "<body><div id='example'><span>EXAMPLE TEXT</span></div></body>"}
```

Templates are composable, so they are allowed to call other templates. The
only catch is that it's not possible to pass an assign to another template
during precompilation. The reason for this is that the logic in a template is
executed the moment the template is called, so if you would pass an assign
during precompilation, the logic in a template would receive this assign
instead of its result, which is only available during runtime. This all means
that when you for example want to pass an assign to a nested template, the
template should be prefixed with the `&` operator, or in other words, executed
during runtime.

```elixir
template templ1,
num: &(&1 + &1) do
  div @num
end

template templ2 do
 h2 @title
 templ1(num: @number) # THIS GENERATES A COMPILE TIME ERROR
 &templ1(num: @number) # THIS IS OK
end
```

Note that because the body of a template is evaluated at compile time, it's
not possible to call other functions from the same module without using `&`
operator.

Instead of defining a do block, you can also provide a path to a file with the
`:file` option.

```elixir
iex> File.write! "test.eml.exs", "div @number"
iex> defmodule MyTemplates3 do
...>   use Eml
...>   use Eml.HTML
...>
...>   template from_file, file: "test.eml.exs"
...> end
iex> File.rm! "test.eml.exs"
iex> MyTemplates3.from_file number: 42
{:safe, "<div>42</div>"}
```

#### Components and fragments

Eml also provides `component/3` and `fragment/3` macros for defining
template elements. They behave as normal elements, but they
aditionally contain a template function that gets called with the
element's attributes and content as arguments during compiling.

```elixir
iex> use Eml
iex> use Eml.HTML
iex> defmodule ElTest do
...>
...>   component my_list,
...>   __CONTENT__: fn content ->
...>     for item <- content do
...>       li do
...>         span "* "
...>         span item
...>         span " *"
...>       end
...>     end
...>   end do
...>     ul [class: @class], @__CONTENT__
...>   end
...>
...> end
iex> import ElTest
iex> el = my_list class: "some-class" do
...>   "Item 1"
...>   "Item 2"
...> end
#my_list<%{class: "some-class"} ["Item 1", "Item 2"]>
iex> Eml.compile(el)
"<ul class='some-class'><li><span>* </span><span>Item 1</span><span> *</span></li><li><span>* </span><span>Item 2</span><span> *</span></li></ul>"
```

Just like templates, its body gets precompiled and assigns, assign
handlers and function calls prefixed with the `operator` are evaluated
at runtime. All attributes of the element can be accessed as assigns
and the element contents is accessable as the assign `@__CONTENT__`.

The main difference between templates and components is their
interface. You can use components like normal elements, even within a
match.

In addition to components, Eml also provides fragments. The difference
between components and fragments is that fragments are without any
logic, so assign handlers or the `&` capture operator are not allowed
in fragment definitions.

Fragments can be used for better composability and performance,
because unlike templates and components, assigns are allowed as
arguments during precompilation for fragments. This is possible
because fragments don't contain any logic.


#### Unpacking

In order to easily access the contents of elements, Eml provides `unpack/1`.
```elixir
iex> Eml.unpack div 42
42

iex> Eml.unpack div span(42)
42
```


#### Querying eml

`Eml.Element` implements the Elixir `Enumerable` protocol for traversing a tree of
nodes. Let's start with creating something to query
```elixir
iex> e = html do
...>   head class: "head" do
...>     meta charset: "UTF-8"
...>   end
...>   body do
...>     article id: "main-content" do
...>       section class: "article" do
...>         h3 "Hello world"
...>       end
...>       section class: "article" do
...>         "TODO"
...>       end
...>     end
...>   end
...> end
#html<[#head<%{class: "head"} [#meta<%{charset: "UTF-8"}>]>,
 #body<[#article<%{id: "main-content"}
  [#section<%{class: "article"} [#h3<"Hello world">]>,
   #section<%{class: "article"} ["TODO"]>]>]>]>
```
To get an idea how the tree is traversed, first just print all nodes
```elixir
iex> Enum.each(e, fn x -> IO.puts(inspect x) end)
#html<[#head<%{class: "head"} [#meta<%{charset: "UTF-8"}>]>, #body<[#article<%{id: "main-content"} [#section<%{class: "article"} [#h3<"Hello world">]>, #section<%{class: "article"} ["TODO"]>]>]>]>
#head<%{class: "head"} [#meta<%{charset: "UTF-8"}>]>
#meta<%{charset: "UTF-8"}>
#body<[#article<%{id: "main-content"} [#section<%{class: "article"} [#h3<"Hello world">]>, #section<%{class: "article"} ["TODO"]>]>]>
#article<%{id: "main-content"} [#section<%{class: "article"} [#h3<"Hello world">]>, #section<%{class: "article"} ["TODO"]>]>
#section<%{class: "article"} [#h3<"Hello world">]>
#h3<"Hello world">
"Hello world"
#section<%{class: "article"} ["TODO"]>
"TODO"
:ok
```

As you can see every node of the tree is passed to `Enum`. Let's continue with
some other examples.
```elixir 
iex> Enum.member?(e, "TODO") true

iex> Enum.filter(e, &Eml.match?(&1, tag: :h3))
[#h3<"Hello world">]

iex> Enum.filter(e, Eml.match?(&1, attrs: %{class: "article"}))
[#section<%{class: "article"} [#h3<"Hello world">]>,

CHANGELOG.md
# Changelog

## v0.9.0-dev
 * Enhancements
  * Replaced quoted expressions with assign handlers and runtime function calls
  * `use Eml` can now be used to set compile options and elements to import
  * Added `Eml.collect/3`
  * Added `Eml.Query`, a small dsl for transforming and collecting Eml nodes

 * Bug fixed
  * Fixed typo in Eml.Element.apply_template/1
  * Suppress new underscore variable compile warning

 * Backwards incompatible changes
  * Quoted expressions are now invalid. Use assign handlers or runtime function calls instead.

## v0.8.0-dev
 * Enhancements
  * Much richer templates by using quoted expressions as replacement for parameters
  * Removed all generic functionality from the html parser and renderer,
    which makes it easier to implement other parsers and renderers
  * Added `{ :safe, String.t }` as a new content type which you can use when you need to add content to an element that should not get escaped
  * Added `transform` option to `Eml.render/3`
  * Added `casing` option to `Eml.Element.Generator.__using__` to control the casing of tags for generated elements
  * Added the `&` capture operator as shortcut for quoted expressions
  * Introduced components and fragments
  * Added `Eml.match?/2` macro
  * Added `Eml.Element.put_template/3`, `Eml.Element.remove_template/1` and `Eml.apply_template/1` functions
  * Added `any` element as a catch all tag macro to be used in a match
  * Added `Eml.escape/1` and `Eml.unescape/1` functions that recursively escape or unescape content.
  * `Eml.Compiler.compile` returns results by default as `{ :safe, result }` so that
    they can be easily added to other elements, witout getting escaped

 * Bug fixes
  * Using element macro's in a match had different confusing behaviour

 * Backwards incompatible changes
  * Removed `Eml.Template` and `Eml.Parameter` in favor of quoted expressions
  * Replaced `Eml.precompile` with `Eml.template` and `Eml.template_fn`
  * Changed names of render options: :lang => :renderer, :quote => :quotes
  * Importing all HTML element macro's is now done via `use Eml.HTML` instead of `use Eml.Language.HTML`
  * Renamed `Eml.Data` protocol to `Eml.Encoder` and removed `Eml.to_content`
  * Data conversion is now only done during compiling and not when adding data to elements.
  * Removed query functions
  * Removed transform functions
  * Removed `defeml` and `defhtml` macros
  * Removed `Eml.unpackr/1` and `Eml.funpackr/1`. `Eml.unpack` now always unpacks recursively
  * Removed `Eml.element?/1`, `Eml.empty?/1` and `Eml.type/1` functions.
  * Removed all previous helper functions from `Eml.Element`
  * Removed `Eml.compile/2` in favor of `Eml.Compiler.compile/2`
  * Parser doesn't automatically converts entities anymore. Use `Eml.unescape/1` instead.

## v0.7.1
 * Enhancements
  * Added unit tests that test escaping and enity parsing
  * Documentation additions and corrections

 * Bug fixes
  * Single and double quotes in attributes now should get properly escaped

 * Backwards incompatible changes

## v0.7.0
 * Enhancements
  * It's now easy to provide conversions for custom data types by using the new `Eml.Data` protocol
  * Better separation of concerns by removing all data conversions from parsing

 * Bug fixes
  * Some type fixes in the Eml.Language behaviour

 * Backwards incompatible changes
  * Renamed `Eml.Language.Html` to `Eml.Language.HTML` in order to be compliant with Elixir's naming conventions
  * The undocumented `Eml.parse/4` function is now replaced by `Eml.to_content/3`
  * The `Eml.Parsable` protocol is replaced by `Eml.Data`, which is now strictly used for converting various
    data types into valid Eml nodes.
  * `Eml.parse/2` now always returns a list again, because the type
    conversions are now done by `Eml.to_content/3` and consequently you can't force
    `Eml.parse/2` to return a list anymore, which would make it dangerous to
    use when parsing html partials where you don't know the nummer of nodes.
  * `Eml.parse/2`, `Eml.render/3` and `Eml.compile/3` now always raise an exception on error.
     Removed `Eml.parse!/2`, `Eml.render!/3` and `Eml.compile!/3`. Reason is that it was hard
     to guarantee that those functions never raised an error and it simplifies Eml's API
  * Removed `Eml.render_to_eex/3`, `Eml.render_to_eex!/3`, `Eml.compile_to_eex/3` and `Eml.compile_to_eex!/3`,
    as they didn't provide much usefulness

## v0.6.0

 * Enhancements
   * Introduced `use Eml.Language.Html` as prefered way of defining markup
   * Restructured README.md and added new content about precompiling
   * It's now possible to pass content as the first argument of an element macro, ie. `div "Hello world!"`
   * Added `Eml.compile_to_eex/3` and `Eml.render_to_eex/3`

 * Bug fixes
  * Documentation corrections.
  * Removed duplicate code in `Eml.defhtml/2`
  * Type specification fixes
  * `Eml.parse/2` in some cases returned weird results when the input is a list
  * Template bindings were not correctly parsed

 * Backwards incompatible changes
  * Removed `Eml.eml/2` in favor of `use Eml.Language.Html`
  * `Eml.parse/2` now returns results in the form of `{ :ok, res }`, in order to be consistent with render and compile functions
  * Unless the input of `Eml.parse/2` is a list, if the result is a single element, `Eml.parse/2` now just returns the single element
    instead of always wrapping the result in a list


## v0.5.0

 * Enhancements
  * Added documentation for all public modules
  * Added some meta data to mix.exs for ex_doc
  * Added `Eml.compile!/3`

 * Bug fixes
  * Lots of documentation corrections
  * Some type fixes

lib/eml.ex
defmodule Eml do
  @moduledoc """
  Eml makes markup a first class citizen in Elixir. It provides a
  flexible and modular toolkit for generating, parsing and
  manipulating markup. It's main focus is html, but other markup
  languages could be implemented as well.

  To start off:

  This piece of code
  ```elixir
  use Eml.HTML

  name = "Vincent"
  age  = 36

  div class: "person" do
    div do
      span "name: "
      span name
    end
    div do
      span "age: "
      span age
    end
  end |> Eml.compile
  ```

  produces
  ```html
  <div class='person'>
    <div>
      <span>name: </span>
      <span>Vincent</span>
    </div>
    <div>
      <span>age: </span>
      <span>36</span>
    </div>
  </div>
  ```

  The functions and macro's in the `Eml` module cover most of
  Eml's public API.
  """

  alias Eml.Element

  @default_elements Eml.HTML
  @default_parser Eml.HTML.Parser

  @type t :: Eml.Encoder.t | [Eml.Encoder.t] | [t]
  @type node_primitive :: String.t | { :safe, String.t } | Macro.t | Eml.Element.t

  @doc """
  Define a template function that compiles eml to a string during compile time.

  Eml uses the assigns extension from `EEx` for parameterizing templates. See
  the `EEx` docs for more info about them. The function that the template macro
  defines accepts optionally any Dict compatible dictionary as argument for
  binding values to assigns.

  ### Example:

      iex> defmodule MyTemplates1 do
      ...>   use Eml
      ...>   use Eml.HTML
      ...>
      ...>   template example do
      ...>     div id: "example" do
      ...>       span @text
      ...>     end
      ...>   end
      ...> end
      iex> MyTemplates.example text: "Example text"
      {:safe, "<div id='example'><span>Example text</span></div>"}


  Eml templates provides two ways of executing logic during runtime. By
  providing assigns handlers to the optional `funs` dictionary, or by calling
  external functions during runtime with the `&` operator.

  ### Example:

      iex> defmodule MyTemplates2 do
      ...>   use Eml
      ...>   use Eml.HTML
      ...>
      ...>   template assigns_handler,
      ...>   text: &String.upcase/1 do
      ...>     div id: "example1" do
      ...>       span @text
      ...>     end
      ...>   end
      ...>
      ...>   template external_call do
      ...>     body &assigns_handler(text: @example_text)
      ...>   end
      ...> end
      iex> MyTemplates.assigns_handler text: "Example text"
      {:safe, "<div id='example'><span>EXAMPLE TEXT</span></div>"}
      iex> MyTemplates.exernal_call example_text: "Example text"
      {:safe, "<body><div id='example'><span>EXAMPLE TEXT</span></div></body>"}


  Templates are composable, so they are allowed to call other templates. The
  only catch is that it's not possible to pass an assign to another template
  during precompilation. The reason for this is that the logic in a template is
  executed the moment the template is called, so if you would pass an assign
  during precompilation, the logic in a template would receive this assign
  instead of its result, which is only available during runtime. This all means
  that when you for example want to pass an assign to a nested template, the
  template should be prefixed with the `&` operator, or in other words, executed
  during runtime.

  ### Example

      iex> defmodule T1 do
      ...>   template templ1,
      ...>   num: &(&1 + &1) do
      ...>     div @num
      ...>   end
      ...> end

      iex> template templ2 do
      ...>   h2 @title
      ...>   templ1(num: @number) # THIS GENERATES A COMPILE TIME ERROR
      ...>   &templ1(num: @number) # THIS IS OK
      ...> end

  Note that because the body of a template is evaluated at compiletime, it's
  not possible to call other functions from the same module without using `&`
  operator.

  Instead of defining a do block, you can also provide a path to a file with the
  `:file` option.

  ### Example:

      iex> File.write! "test.eml.exs", "div @number"
      iex> defmodule MyTemplates3 do
      ...>   use Eml
      ...>   use Eml.HTML
      ...>
      ...>   template from_file, file: "test.eml.exs"
      ...> end
      iex> File.rm! "test.eml.exs"
      iex> MyTemplates3.from_file number: 42
      {:safe, "<div>42</div>"}

  """
  defmacro template(name, funs \\ [], do_block) do
    do_template(name, funs, do_block, __CALLER__, false)
  end

  @doc """
  Define a private template.

  Same as `template/3` except that it defines a private function.
  """
  defmacro templatep(name, funs \\ [], do_block) do
    do_template(name, funs, do_block, __CALLER__, true)
  end

  defp do_template(tag, funs, do_block, caller, private) do
    { tag, _, _ } = tag
    def_call = if private, do: :defp, else: :def
    template = Eml.Compiler.precompile(caller, do_block)
    quote do
      unquote(def_call)(unquote(tag)(var!(assigns))) do
        _ = var!(assigns)
        var!(funs) = unquote(funs)
        _ = var!(funs)
        unquote(template)
      end
    end
  end

  @doc """
  Define a template as an anonymous function.

  Same as `template/3`, except that it defines an anonymous function.

  ### Example
      iex> t = template_fn names: fn names ->
      ...>   for n <- names, do: li n
      ...> end do
      ...>   ul @names
      ...> end
      iex> t.(names: ~w(john james jesse))
      {:safe, "<ul><li>john</li><li>james</li><li>jesse</li></ul>"}

  """
  defmacro template_fn(funs \\ [], do_block) do
    template = Eml.Compiler.precompile(__CALLER__, do_block)
    quote do
      fn var!(assigns) ->
        _ = var!(assigns)
        var!(funs) = unquote(funs)
        _ = var!(funs)
        unquote(template)
      end
    end
  end

  @doc """
  Define a component element

  Components in Eml are a special kind of element that inherit functionality
  from templates. Like templates, everything within the do block gets
  precompiled, except assigns and function calls prefixed with the `&`
  operator. Defined attributes on a component can be accessed as assigns, just
  like with templates. Content can be accessed via the the special assign
  `__CONTENT__`. However, since the type of a component is `Eml.Element.t`,
  they can be queried and transformed, just like normal Eml elements.

  See `template/3` for more info about composability, assigns, runtime logic and
  accepted options.

  ### Example

      iex> use Eml
      iex> use Eml.HTML
      iex> defmodule ElTest do
      ...>
      ...>   component my_list,
      ...>   __CONTENT__: fn content ->
      ...>     for item <- content do
      ...>       li do
      ...>         span "* "
      ...>         span item
      ...>         span " *"
      ...>       end
      ...>     end
      ...>   end do
      ...>     ul [class: @class], @__CONTENT__
      ...>   end
      ...>
      ...> end
      iex> import ElTest
      iex> el = my_list class: "some-class" do
      ...>   "Item 1"
      ...>   "Item 2"
      ...> end
      #my_list<%{class: "some-class"} ["Item 1", "Item 2"]>
      iex> Eml.compile(el)
      "<ul class='some-class'><li><span>* </span><span>Item 1</span><span> *</span></li><li><span>* </span><span>Item 2</span><span> *</span></li></ul>"
  """
  defmacro component(tag, funs \\ [], do_block) do
    do_template_element(tag, funs, do_block, __CALLER__, false)
  end

  @doc """
  Define a fragment element

  Fragments in Eml are a special kind of element that inherit functionality from
  templates. Like templates, everything within the do block gets precompiled,
  except assigns. Defined attributes on a component can be accessed as assigns,
  just like with templates. Content can be accessed via the the special assign
  `__CONTENT__`.  However, since the type of a fragment is `Eml.Element.t`, they
  can be queried and transformed, just like normal Eml elements.

  The difference between components and fragments is that fragments are without
  any logic, so assign handlers or the `&` operator are not allowed in a
  fragment definition.

  The reason for their existence is easier composability and performance,
  because unlike templates and components, it is allowed to pass assigns to
  fragments during precompilation. This is possible because fragments don't
  contain any logic.

  ### Example

      iex> use Eml
      nil
      iex> use Eml.HTML
      nil
      iex> defmodule ElTest do
      ...>
      ...>   fragment basic_page do
      ...>     html do
      ...>       head do
      ...>         meta charset: "UTF-8"
      ...>         title @title
      ...>       end
      ...>       body do
      ...>         @__CONTENT__
      ...>       end
      ...>     end
      ...>   end
      ...>
      ...> end
      {:module, ElTest, ...}
      iex> import ElTest
      nil
      iex> page = basic_page title: "Hello!" do
      ...>   div "Hello World"
      ...> end
      #basic_page<%{title: "Hello!!"} [#div<"Hello World">]>
      iex> Eml.compile page
      "<!doctype html>\n<html><head><meta charset='UTF-8'/><title>Hello!!</title></head><body><div>Hello World</div></body></html>"
  """
  defmacro fragment(tag, do_block) do
    do_template_element(tag, nil, do_block, __CALLER__, true)
  end

  defp do_template_element(tag, funs, do_block, caller, fragment?) do
    { tag, _, _ } = tag
    template = Eml.Compiler.precompile(caller, Keyword.merge(do_block, fragment: fragment?))
    template_tag = (Atom.to_string(tag) <> "__template") |> String.to_atom()
    template_type = if fragment?, do: :fragment, else: :component
    funs = unless fragment? do
      quote do
        var!(funs) = unquote(funs)
        _ = var!(funs)
      end
    end
    quote do
      @doc false
      def unquote(template_tag)(var!(assigns)) do
        _ = var!(assigns)
        unquote(funs)
        unquote(template)
      end
      defmacro unquote(tag)(content_or_attrs, maybe_content \\ nil) do
        tag = unquote(tag)
        template_tag = unquote(template_tag)
        template_type = unquote(template_type)
        in_match = Macro.Env.in_match?(__CALLER__)
        { attrs, content } = Eml.Element.Generator.extract_content(content_or_attrs, maybe_content, in_match)
        if in_match do
          quote do
            %Eml.Element{tag: unquote(tag), attrs: unquote(attrs), content: unquote(content)}
          end
        else
          quote do
            %Eml.Element{tag: unquote(tag),
                         attrs: Enum.into(unquote(attrs), %{}),
                         content: List.wrap(unquote(content)),
                         template: &unquote(__MODULE__).unquote(template_tag)/1,
                         type: unquote(template_type)}
          end
        end
      end
    end
  end

  @doc """
  Parses data and converts it to eml

  How the data is interpreted depends on the `parser` argument.
  The default value is `Eml.HTML.Parser', which means that
  strings are parsed as html.

  In case of error, raises an Eml.ParseError exception.

  ### Examples:

      iex> Eml.parse("<body><h1 id='main-title'>The title</h1></body>")
      [#body<[#h1<%{id: "main-title"} "The title">]>]
  """
  @spec parse(String.t, Keyword.t) :: [t]
  def parse(data, opts \\ [])

  def parse(data, opts) when is_binary(data) do
    parser = opts[:parser] || @default_parser
    parser.parse(data, opts)
  end
  def parse(data, _) do
    raise Eml.ParseError, message: "Bad argument: #{inspect data}"
  end

  @doc """
  Compiles eml content with the specified markup compiler, which is html by default.

  The accepted options are:

  * `:compiler` - The compiler to use, by default `Eml.HTML.Compiler`
  * `:quotes` - The type of quotes used for attribute values. Accepted values are `:single` (default) and `:double`.
  * `:transform` - A function that receives every node just before it get's compiled. Same as using `transform/2`,
     but more efficient, since it's getting called during the compile pass.
  * `:escape` - Automatically escape strings, default is `true`.

  In case of error, raises an Eml.CompileError exception.

  ### Examples:

      iex> Eml.compile(body(h1([id: "main-title"], "A title")))
      "<body><h1 id='main-title'>A title</h1></body>"

      iex> Eml.compile(body(h1([id: "main-title"], "A title")), quotes: :double)
      "<body><h1 id=\"main-title\">A title</h1></body>"

      iex> Eml.compile(p "Tom & Jerry")
      "<p>Tom &amp; Jerry</p>"

  """
  @spec compile(t, Dict.t) :: String.t
  def compile(content, opts \\ [])
  def compile({ :safe, string }, _opts) do
    string
  end
  def compile(content, opts) do
    case Eml.Compiler.compile(content, opts) do
      { :safe, string } ->
        string
      _ ->
        raise Eml.CompileError, message: "Bad argument: #{inspect content}"
    end
  end

  @doc """
  Recursively transforms `eml` content.

  It traverses all nodes of the provided eml tree.  The provided transform
  function will be evaluated for every node `transform/3` encounters. Parent
  nodes will be transformed before their children. Child nodes of a parent will
  be evaluated before moving to the next sibling.

  When the provided function returns `nil`, the node will be removed from the
  eml tree.

  Note that because parent nodes are evaluated before their children, no
  children will be evaluated if the parent is removed.

  ### Examples:

      iex> e = div do
      ...>   span [id: "inner1", class: "inner"], "hello "
      ...>   span [id: "inner2", class: "inner"], "world"
      ...> end
      #div<[#span<%{id: "inner1", class: "inner"} "hello ">,
       #span<%{id: "inner2", class: "inner"} "world">]>

      iex> Eml.transform(e, fn
      ...>   span(_) -> "matched"
      ...>   node    -> node
      ...> end)
      #div<["matched", "matched"]>

      iex> transform(e, fn node ->
      ...>   IO.inspect(node)
      ...>   node
      ...> end)
      #div<[#span<%{class: "inner", id: "inner1"} "hello ">,
       #span<%{class: "inner", id: "inner2"} "world">]>
      #span<%{class: "inner", id: "inner1"} "hello ">
      "hello "
      #span<%{class: "inner", id: "inner2"} "world">
      "world"
      #div<[#span<%{class: "inner", id: "inner1"} "hello ">,
       #span<%{class: "inner", id: "inner2"} "world">]>
  """
  @spec transform(t, (t -> t)) :: t | nil
  def transform(nil, _fun) do
    nil
  end
  def transform(eml, fun) when is_list(eml) do
    for node <- eml, t = transform(node, fun), do: t
  end
  def transform(node, fun) do
    case fun.(node) do
      %Element{content: content} = node ->
        %Element{node| content: transform(content, fun)}
      node ->
        node
    end
  end

  @doc """
  Recursively reduces a tree of nodes

  ### Example

      iex> tree = div do
      ...>   span 1
      ...>   span 2
      ...> end
      iex> list = [tree, tree]

      iex> Eml.collect(list, [], fn node, acc ->
      ...>   if is_integer(node) do
      ...>     [node | acc]
      ...>   else
      ...>     acc
      ...>   end
      ...> end)
      [2, 1, 2, 1]
  """
  @spec collect(t, term, (t, term -> term)) :: term
  def collect(eml, acc \\ %{}, fun)
  def collect(eml, acc, fun) when is_list(eml) do
    Enum.reduce(eml, acc, &collect(&1, &2, fun))
  end
  def collect(%Element{} = eml, acc, fun) do
    Enum.reduce(eml, acc, fun)
  end
  def collect(eml, acc, fun) do
    fun.(eml, acc)
  end


  @doc """
  Match on element tag, attributes, or content

  Implemented as a macro.
  ### Examples:

      iex> use Eml
      iex> use Eml.HTML
      iex> node = section [id: "my-section"], [div([id: "some_id"], "Some content"), div([id: "another_id"], "Other content")]
      iex> Eml.match?(node, attrs: %{id: "my-section"})
      true
      iex> Eml.match?(node, tag: :div)
      false
      iex> Enum.filter(node, &Eml.match?(&1, tag: :div))
      [#div<%{id: "some_id"} "Some content">, #div<%{id: "another_id"}
      "Other content">]
      iex> Eml.transform(node, fn node ->
      ...>   if Eml.match?(node, content: "Other content") do
      ...>     put_in(node.content, "New content")
      ...>   else
      ...>     node
      ...>   end
      ...> end)
      #section<%{id: "my-section"}
      [#div<%{id: "some_id"} "Some content">, #div<%{id: "another_id"}
       "New content">]>
  """
  defmacro match?(node, opts \\ []) do
    tag     = opts[:tag]     || quote do: _
    attrs   = opts[:attrs]   || quote do: _
    content = opts[:content] || quote do: _
    quote do
      case unquote(node) do
        %Eml.Element{tag: unquote(tag), attrs: unquote(attrs), content: unquote(content)} ->
          true
        _ ->
          false
      end
    end
  end

  @doc """
  Extracts a value recursively from content

  ### Examples

      iex> Eml.unpack [42]
     bm 42

      iex> Eml.unpack 42
      42

      iex> Eml.unpack(div "hallo")
      "hallo"

      iex> Eml.unpack Eml.unpack(div(span("hallo")))
      "hallo"

      iex> Eml.unpack div(span(42))
      42

      iex> Eml.unpack div([span("Hallo"), span(" world")])
      ["Hallo", " world"]

  """
  @spec unpack(t) :: t
  def unpack(%Element{content: content}) do
    unpack(content)
  end
  def unpack([node]) do
    unpack(node)
  end
  def unpack(content) when is_list(content) do
    for node <- content, do: unpack(node)
  end
  def unpack({ :safe, node }) do
    node
  end
  def unpack(node) do
    node
  end

  @doc """
  Escape content

  ### Examples

      iex> escape "Tom & Jerry"
      "Tom &amp; Jerry"
      iex> escape div span("Tom & Jerry")
      #div<[#span<["Tom &amp; Jerry"]>]>
  """
  @spec escape(t) :: t
  defdelegate escape(eml), to: Eml.Compiler

  @doc """
  Unescape content

  ### Examples

      iex> unescape "Tom &amp; Jerry"
      "Tom & Jerry"
      iex> unescape div span("Tom &amp; Jerry")
      #div<[#span<["Tom & Jerry"]>]>
  """
  @spec unescape(t) :: t
  defdelegate unescape(eml), to: Eml.Parser

  # use Eml
  @doc """
  Import macro's from this module and alias `Eml.Element`.

  Accepts the following options:

  * `:compile` - Set dcompile options as a Keyword list for all templates,
    components and fragments that are defined in the module where `use Eml` is
    invoked. See `Eml.compile/2` for all available options.
  * `:elements` - Which elements to import in the current scope. Accepts a
    module, or list of modules and defaults to `Eml.HTML`. When you don't want
    to import any elements, set to `nil` or `false`.
  """
  defmacro __using__(opts) do
    use_elements = if mods = Keyword.get(opts, :elements, @default_elements) do
                 for mod <- List.wrap(mods) do
                   quote do: use unquote(mod)
                 end
               end
    compile_opts = Keyword.get(opts, :compile, [])
    if mod = __CALLER__.module do
      Module.put_attribute(mod, :eml_compile, compile_opts)
    end
    quote do
      unquote(use_elements)
      import Eml, only: [
        template: 2, template: 3,
        templatep: 2, templatep: 3,
        template_fn: 1, template_fn: 2,
        component: 2, component: 3,
        fragment: 2
      ]
    end
  end
end
